### PR TITLE
ENH: Improve FIR path of signal.decimate

### PR DESCRIPTION
**Pull request by e-q**
This PR aims to relieve the phase shift due to the group delay of an FIR downsampling filter in signal.decimate through the use of `resample_poly`.

This requires modifying `resample_poly` to accept arbitrary FIR filter coefficients, as `decimate` itself advertises that capability. A test for `resample_poly` has been added for this capability. Hopefully this isn't too much of a hack. (Thoughts, @Eric89GXL ?)

### CODE REVIEW

**Comment by Eric89GXL** [comment]
> This requires modifying resample_poly to accept arbitrary FIR filter coefficients

It's good to add that capability. I didn't do it originally to keep the PR simple, and to offload the work of getting it right on the next developer. Looks like you fell into my trap :)

File: scipy/signal/signaltools.py [hunk 1]
```diff
@@ -1896,6 +1896,9 @@ def resample_poly(x, up, down, axis=0, window=('kaiser', 5.0)):
         Desired window to use to design the low-pass filter. See
         `scipy.signal.get_window` for a list of windows and required
         parameters.
+    num : array_like, optional
+         FIR filter coefficients of the low-pass filter to be used. Overrides
+         the `window` argument if not None.
```
Comment (Eric89GXL) [inline scipy/signal/signaltools.py]: Just roll this into `window`. See e.g. how `resample` allows multiple types for `window`:

http://docs.scipy.org/doc/scipy-0.16.0/reference/generated/scipy.signal.resample.html

File: scipy/signal/signaltools.py [hunk 2]
```diff
@@ -2888,9 +2895,10 @@ def decimate(x, q, n=None, ftype='iir', axis=-1, zero_phase=None):
         The axis along which to decimate.
     zero_phase : bool, optional
         Prevent phase shift by filtering with `filtfilt` instead of `lfilter`
-        when `ftype` is not 'fir'.  A value of `True` is recommended, since a
-        phase shift is generally not desired. Using `None` defaults to `False`
-        for backwards compatibility.
+        when using an IIR filter, and shifting the outputs back by the filter's
+        group delay when using an FIR filter.  A value of `True` is
```
Comment (Eric89GXL) [inline scipy/signal/signaltools.py]: double-backticks on `True`

File: scipy/signal/signaltools.py [hunk 3]
```diff
@@ -2888,9 +2895,10 @@ def decimate(x, q, n=None, ftype='iir', axis=-1, zero_phase=None):
...
+        recommended, since a phase shift is generally not desired. Using `None`
+        defaults to `False` for backwards compatibility.
```
Comment (Eric89GXL) [inline scipy/signal/signaltools.py]: oh yeah, this default is going to change, right? Better to mention here (in addition to the warning you raise) that the `None` default behavior will change in a subsequent release, so better to set it explicitly.

File: scipy/signal/signaltools.py [hunk 4]
```diff
@@ -2920,35 +2928,47 @@ def decimate(x, q, n=None, ftype='iir', axis=-1, zero_phase=None):
...
 
-    if zero_phase:
-        y = filtfilt(system.num, system.den, x, axis=axis)
-    else:
-        y = lfilter(system.num, system.den, x, axis=axis)
+        if zero_phase:
+            y = resample_poly(x, 1, q, axis=axis, num=system.num)
+            return y
+        else:
+            y = lfilter(system.num, 1.0, x, axis=axis)
```
Comment (Eric89GXL) [inline scipy/signal/signaltools.py]: even here you can use `upfirdn` to save time

**Comment by Eric89GXL** [comment]
So we're swapping in (currently) `resample_poly` for `filtfilt` and (soon) `upfirdn` for `lfilter` to save time. It would be nice in the tests to have the outputs computed using the naive `filtfilt` and `lfilter` algorithms, and compare outputs of `decimate` (which will use `resample_poly` and `upfirdn` under the hood). They should hopefully match well using `assert_allclose`. It helps ensure we are really getting equivalent operations. WDYT?

**Comment by Eric89GXL** [comment]
BTW can you add a comment about the speed gains? We doing `1 / down` operations by using `upfirdn` compared to `lfilter`. Also, something neat under the hood is that instead of doing a two-pass operation for `filtfilt`, `resample_poly` does a single filter operation by taking advantage of associativity `(x(t) * h(t)) * h(-t) = (x(t) * (h(t) * h(-t)))`, i.e. convolving the filter with it's time-reversed version, then dealing with the zero padding properly.

Speaking of which, this is one way that you might need to change the code -- I think that `filtfilt` would apply the same filter twice -- once forward and once backward -- whereas `resample_poly` will only apply it once (and it will assume you've constructed your filter in this symmetrical way). So maybe in this function, the `window` function/array needs to be convolved with it's time-reversed version before being passed to `resample_poly`? You should see it in the tests if you implement what I mention above.

**Comment by e-q** [comment]
Thanks for the feedback @Eric89GXL! I've incorporated much of your feedback.

The speed gains of `upfirdn` are quite good for larger downsampling factors, which is great. Where were you thinking of the speedups being mentioned, besides the commit message?

As for the `upfirdn` vs. `lfilter` results, I'm opening a separate PR that adds it to the `upfirdn` tests (which seems a little more natural to me). I haven't yet tried `resample_poly` vs. `filtfilt` yet. I think this should work as long as the `filtfilt` is explicitly given the zero-padding argument...

**Comment by Eric89GXL** [comment]
> Where were you thinking of the speedups being mentioned, besides the commit message?

A comment in the code saying why `upfirdn` is preferred to `lfilter` (polyphase resampling avoids unnecessary calculations) would be a good place

> I haven't yet tried resample_poly vs. filtfilt yet. I think this should work as long as the filtfilt is explicitly given the zero-padding argument...

I don't expect the zero-padding to really matter. `resample_poly` must zero pad and remove zeros at the end because it only does a single pass, in one direction, and computes only specific subsamples -- this introduces a phase shift, so the data are not centered after the operation. `filtfilt` on the other hand does two passes, one forward and one backward, and this is a zero-phase operation. So hopefully you don't have to do anything too special, mostly:

1. Pass `h` as the filter to `filtfilt`
2. Extend `resample_poly` to allow for arbitrary filters
3. Pass `convolve(h, h[::-1])` as the filter to `resample_poly`

**Comment by e-q** [comment]
Ok, testing vs `filtfilt` is included now. You were exactly right about passing the convolved filter, thanks.

I was unclear before, the zero-padding I was referring to was in regards to the input signal, not the filter. `filtfilt` does an odd signal extension by default, but can do constant padding. Since, according to the docstring, `resample_poly` assumes zeros outside of the input samples, I set the first and last samples to zero and used `filtfilt`'s constant padding mode. Anyways, no big deal, the test passes on my machine.

**Comment by Eric89GXL** [comment]
That makes sense. Glad they match. I'll look deeper soon.

Any practical numbers on what the resample_poly speedup is like over
filtfilt? I'm curious how close it gets to being `down` times faster.

**Comment by e-q** [comment]
Decimating x=np.random.randn(10**7) with `resample_poly`, using a 31 tap FIR filter, is actually faster than `filtfilt` by more than a factor of `down` on my machine.
```
q    filtfilt    resample_poly   speedup
----------------------------------------
2    591ms       211.0ms          2.8
5    577ms        98.7ms          5.8
13   604ms        44.4ms         13.6
```
This comes from the additional time cost of slicing the output of `filtfilt`. Without the slicing, the speedup for `q=13` is 12.97, for instance.

**Comment by Eric89GXL** [comment]
Excellent. Is this good to go from your end, then?

**Comment by e-q** [comment]
yep!

File: scipy/signal/signaltools.py [hunk 5]
```diff
@@ -1959,13 +1966,20 @@ def resample_poly(x, up, down, axis=0, window=('kaiser', 5.0)):
...
     if up == down == 1:
         return x.copy()
-    n_out = (x.shape[axis] * up) // down
-
-    # Design a linear-phase low-pass FIR filter
-    max_rate = max(up, down)
-    f_c = 1. / max_rate  # cutoff of FIR filter (rel. to Nyquist)
-    half_len = 10 * max_rate  # reasonable cutoff for our sinc-like function
-    h = firwin(2 * half_len + 1, f_c, window=window)
+    n_out = int(np.ceil(x.shape[axis] * up / down))
```
Comment (Eric89GXL) [inline scipy/signal/signaltools.py]: This used to effectively be `floor`, why the change to `ceil`?

I assume it's for the extra sample in case it doesn't divide evenly, In that case it would actually be better to use integer arithmetic if possible so we stay exact, like:
```
n_out = x.shape['axis'] * up
n_out = n_out // down + bool(n_out % down)
```

File: scipy/signal/signaltools.py [hunk 6]
```diff
@@ -1917,11 +1916,19 @@ def resample_poly(x, up, down, axis=0, window=('kaiser', 5.0)):
...
     the filter length and `down` (see `scipy.signal.upfirdn` for details).
 
-    The `window` argument is passed directly to `scipy.signal.firwin`
-    to design a low-pass filter.
+    The argument `window` specifies the FIR low-pass filter design.
+
+    If `window` is an array_like it is assumed to be the FIR filter
+    coefficients. Note that the FIR filter is applied after the upsampling
+    step, so it should be designed to operate on a signal at a sampling
+    frequency higher than the original by a factor of `up//gcd(up, down)`.
```
Comment (Eric89GXL) [inline scipy/signal/signaltools.py]: We should probably also mention that the output will be centered with respect to this array, so it might be best to pass a symmetric filter with an odd number of samples to ensure a zero-phase filter.

File: scipy/signal/signaltools.py [hunk 7]
```diff
@@ -2920,35 +2937,44 @@ def decimate(x, q, n=None, ftype='iir', axis=-1, zero_phase=None):
...
+            y = filtfilt(system.num, system.den, x, axis=axis)
+        else:
+            y = lfilter(system.num, system.den, x, axis=axis)
 
-    sl = [slice(None)] * y.ndim
-    sl[axis] = slice(None, None, q)
-    return y[sl]
+        sl = [slice(None)] * y.ndim
+        sl[axis] = slice(None, None, q)
+        return y[sl]
```
Comment (Eric89GXL) [inline scipy/signal/signaltools.py]: Instead of having four different `return` paths, I usually prefer having one `return` at the end if possible, so each of these conditionals would just assign `y`. Thoughts?

**Comment by Eric89GXL** [comment]
Other than my minor comments, LGTM.

**Comment by e-q** [comment]
Good points all around, addressed and squashed. Thanks!

The short answer about the change of `n_out` for `resample_poly` was to make it match the number of outputs from `decimate`, which in turn is given by how many outputs you get from slicing `x[::q]`. 

Incidentally, you'll see in the `zero_phase=False, ftype='fir'` path that the output from `upfirdn` is truncated to that same amount. It seems that `upfirdn` tacks a few more output points on than I would expect from the decimation point of view. Why is this?

**Comment by Eric89GXL** [comment]
IIRC `upfirdn` lets the filter ring out to the end, like `np.convolve(..., mode='full')`

**Comment by e-q** [comment]
Is this good to go?

File: scipy/signal/signaltools.py [hunk 8]
```diff
@@ -1917,11 +1916,22 @@ def resample_poly(x, up, down, axis=0, window=('kaiser', 5.0)):
...
-    The `window` argument is passed directly to `scipy.signal.firwin`
-    to design a low-pass filter.
+    The argument `window` specifies the FIR low-pass filter design.
+
+    If `window` is an array_like it is assumed to be the FIR filter
+    coefficients. Note that the FIR filter is applied after the upsampling
+    step, so it should be designed to operate on a signal at a sampling
+    frequency higher than the original by a factor of `up//gcd(up, down)`.
+    This function's output will be centered with respect to this array, so it
+    is best to pass a symmetric filter with an odd number of samples to if, as
```
Comment (Eric89GXL) [inline scipy/signal/signaltools.py]: "number of samples to if" -> "number of samples if"

**Comment by Eric89GXL** [comment]
Other than my one new gripe, +1 for merge. After the rewording, let's wait a day or two to see if anyone else wants to comment then merge

**Comment by e-q** [comment]
Thanks for the look! The wording has been addressed.

**Comment by Eric89GXL** [comment]
Thanks @e-q

File: benchmarks/benchmarks/signal_filtering.py [hunk 9]
```diff
@@ -3,12 +3,29 @@
 import numpy as np
 
 try:
-    from scipy.signal import lfilter, firwin
+    from scipy.signal import lfilter, firwin, decimate
```
Comment (pv) [inline benchmarks/benchmarks/signal_filtering.py]: Usually best to put these in a separate import statement, so that not having decimate does not prevent running eg the firwin benchmarks
